Merge makefiles in otherlibs/systhreads

shindere opened the pull request:
The initial project was to submit one PR merging all the makefiles in
all subdirectories of otherlibs. However it turned out that just the merge
in the systhreads directory is rather non trivial and may deserve a separate
review.

@alainfrisch you may want to review it since it seems you have worked on it.

shindere committed:
Merge makefiles in otherlibs/systhreads

Before this commit, the .cma and .cmxa libraries were linked using -linkall
under Windows but not under Unix.

It was not possible to clarify why -linkall was useful, but given the small
size of the involved modules and the fact that, for most modules, it
does not seem to make much sense to use one without the others, it has been
decided to use -linkall everywhere.

This commit also stops building the .cmxs shared library under Windows, for
consistency reasons (it was built only under Windows before).

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -14,95 +14,129 @@
+ifeq "$(UNIX_OR_WIN32)" "unix"
+	$(MKLIB) -o $(LIBNAME) -ocamlc '$(CAMLC)' -cclib -lunix $(PTHREAD_CAML_LINK) $^
+else # Windows
+	$(MKLIB) -o $(LIBNAME) -ocamlc "$(CAMLC)" -linkall $^
```
Why is linkall used under Windows and not Unix?

alainfrisch replied:
Merging makefiles has the nice benefit of making such differences apparent, and one should then really either remove or document them.  But indeed, this can come later.

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -14,95 +14,129 @@
+st_stubs_n.$(O): st_stubs.c $(HEADER)
+	$(NATIVECC) -I../../asmrun -I../../byterun $(NATIVECCCOMPOPTS) \
+	  $(SHAREDCCCOMPOPTS) -DNATIVE_CODE -DTARGET_$(ARCH) \
+	  -DMODEL_$(MODEL) -DSYS_$(SYSTEM) -c st_stubs.c
```
I'd suggest using `$<` as in the rule above.  And also `$(ROOTDIR)` instead of  `../..`.

Wouldn't it be better to copy the source file before compiling, instead of moving the resulting object file?  I'm thinking about parallel make.

shindere replied:
Alain Frisch (2016/08/30 00:35 -0700):

> >  # Indeed, if we link threads.cmxa, then we must also link unix.cmxa,
> >  # which itself will pass -lunix to the C linker.  It seems more
> >  # modular to me this way. -- Alain
> > 
> > +$(LIBNAME).cmxs: $(LIBNAME).cmxa lib$(LIBNAME)nat.$(A)
> > -   $(CAMLOPT) -shared -o $@ -I . $< -linkall
> >   +
> >   +st_stubs_b.$(O): st_stubs.c $(HEADER)
> > -   $(BYTECC) -I../../byterun $(BYTECCCOMPOPTS) $(SHAREDCCCOMPOPTS) -c $<
> > -   mv st_stubs.$(O) $@
> >   +
> >   +st_stubs_n.$(O): st_stubs.c $(HEADER)
> > -   $(NATIVECC) -I../../asmrun -I../../byterun $(NATIVECCCOMPOPTS) \
> > -     $(SHAREDCCCOMPOPTS) -DNATIVE_CODE -DTARGET_$(ARCH) \
> > -     -DMODEL_$(MODEL) -DSYS_$(SYSTEM) -c st_stubs.c
> 
> I'd suggest using `$<` as in the rule above.

Done, thanks.

> And also `$(ROOTDIR)` instead of  `../..`.

Done, too. I actually made sure ../.. was not used in any place and
systematically replaced it with $(ROOTDIR).

> Wouldn't it be better to copy the source file before compiling,
> instead of moving the resulting object file?  I'm thinking about
> parallel make.

I noticed the issue, too. Again my plan was to address it later but if
you prefere it can be part of this commit, too.

Regarding the way to address it, the idea I had in mind was not
copy-based. The approach I'd suggest would be to introduce three files:
st_stubs.c for all the common stuff and st_stubs_b.c and st_stubs_n.c
for what is specific to bytecode resp. nativecode. Even if e.g. the
bytecode version just includes the common file. And in the nativecode
version one could #define the macros rather than passing them to the
compiler in the Makefile. What do you think?

Oh and by the way, that was another difference between the build
systems: not the same macros were defined under Unix and Windows.

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -14,95 +14,129 @@
+ifeq "$(UNIX_OR_WIN32)" "unix"
+	$(CAMLOPT) -a -o $@ $^ -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
+else # Windows
+	$(MKLIB) -o $(LIBNAME)nat -ocamlopt "$(CAMLOPT)" -linkall $^ $(LINKOPTS)
```
Wouldn't it work to the same as under unix here (i.e. use ocamlopt directly to build the lib)?

shindere replied:
Alain Frisch (2016/08/30 00:37 -0700):

> >  # See remark above: force static linking of libthreadsnat.a
> > -threads.cmxa: $(THREAD_OBJS:.cmo=.cmx)
> > -   $(CAMLOPT) -a -o threads.cmxa $(THREAD_OBJS:.cmo=.cmx) \
> > -     -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
> >   +$(LIBNAME).cmxa: $(THREADS_NCOBJS)
> >   +ifeq "$(UNIX_OR_WIN32)" "unix"
> > -   $(CAMLOPT) -a -o $@ $^ -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
> >   +else # Windows
> > -   $(MKLIB) -o $(LIBNAME)nat -ocamlopt "$(CAMLOPT)" -linkall $^ $(LINKOPTS)
> 
> Wouldn't it work to the same as under unix here (i.e. use ocamlopt
> directly to build the lib)?

It does indeed, thanks.

So I updated the PR accordingly and also added a TODO regarding the
linkall issue, so I think I have now taken into account all the comments
so far.

shindere commented on otherlibs/systhreads/Makefile:
```diff
@@ -14,95 +14,129 @@
+ifeq "$(UNIX_OR_WIN32)" "unix"
+	$(MKLIB) -o $(LIBNAME) -ocamlc '$(CAMLC)' -cclib -lunix $(PTHREAD_CAML_LINK) $^
+else # Windows
+	$(MKLIB) -o $(LIBNAME) -ocamlc "$(CAMLC)" -linkall $^
```
Alain Frisch (2016/08/30 00:28 -0700):

> > -threads.cma: $(THREAD_OBJS)
> > -   $(MKLIB) -ocamlc '$(CAMLC)' -o threads $(THREAD_OBJS) \
> > -     -cclib -lunix $(PTHREAD_CAML_LINK)
> >   +$(LIBNAME).cma: $(THREADS_BCOBJS)
> >   +ifeq "$(UNIX_OR_WIN32)" "unix"
> > -   $(MKLIB) -o $(LIBNAME) -ocamlc '$(CAMLC)' -cclib -lunix $(PTHREAD_CAML_LINK) $^
> >   +else # Windows
> > -   $(MKLIB) -o $(LIBNAME) -ocamlc "$(CAMLC)" -linkall $^
> 
> Why is linkall used under Windows and not Unix?

To be honest, I don't know. It's just the way things were done before so
I tried to be as conservative as possible for this commit, in order to
separate changes from each other. But if you think it would be better to
add linkall under Linux as well I'm perfectly fine and can certainly do
so, so please just let me know.

Thanks a lot for the review!

shindere replied:
Alain Frisch (2016/08/30 01:08 -0700):

> > -threads.cma: $(THREAD_OBJS)
> > -   $(MKLIB) -ocamlc '$(CAMLC)' -o threads $(THREAD_OBJS) \
> > -     -cclib -lunix $(PTHREAD_CAML_LINK)
> >   +$(LIBNAME).cma: $(THREADS_BCOBJS)
> >   +ifeq "$(UNIX_OR_WIN32)" "unix"
> > -   $(MKLIB) -o $(LIBNAME) -ocamlc '$(CAMLC)' -cclib -lunix $(PTHREAD_CAML_LINK) $^
> >   +else # Windows
> > -   $(MKLIB) -o $(LIBNAME) -ocamlc "$(CAMLC)" -linkall $^
> 
> Merging makefiles has the nice benefit of making such differences
> apparent, and one should then really either remove or document them.
> But indeed, this can come later.

I agree. For this specific example, it's just that it' was not obvious
to me what to do.

There are other places where I got rid of the differences, e.g. IIRC not
all the files were installed properly on Windows, I think it was the
.mli files which were not installed.

In the same vein, the warning-related flags were also different and I
took the most strict ones.

alainfrisch commented:
Generally speaking, I'd suggest to add TODO comments to make it explicit when you discover "unexplained" differences between the two versions during the merge.  Otherwise, people would easily assume that an explicit conditional in a Makefile is on purpose and not consider investigating the reason for the difference later.

shindere commented:
That's a good idea and I will do that, thanks a lot.

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
 
-libthreads.a: $(BYTECODE_C_OBJS)
-	$(MKLIB) -o threads $(BYTECODE_C_OBJS) $(PTHREAD_LINK)
+allopt: lib$(LIBNAME).$(A) $(LIBNAME).cmxa $(LIBNAME).cmxs $(CMIFILES)
```
This should be `lib$(LIBNAME)nat.$(A)`, no?

alainfrisch replied:
Yes, but it was probably wrong (and Makefile was right).

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
 # which itself will pass -lunix to the C linker.  It seems more
 # modular to me this way. -- Alain
 
+$(LIBNAME).cmxs: $(LIBNAME).cmxa lib$(LIBNAME)nat.$(A)
```
The .cmxs file was not built on Unix previously, probably because "Dynamic linking with -lpthread is risky on many platforms" as another comment said.  It used to be built under Windows, though.  I'd rather not starting to build the .cmxs file under Unix, unless you have a good reason to do so.  And I would use the opportunity to stop building it under Windows as well for the sake of uniformity across platforms.

alainfrisch replied:
Simplifying the build system is secondary, I agree.  But even for users, declaring once and for all (and documenting) that the systhreads library does not provide a .cmxs file is better, in my opinion, than providing the .cmxs file only on Windows.

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
-	$(AR) rc libthreadsnat.a $(NATIVECODE_C_OBJS)
+	$(AR) rc $@ $^
+else # Windows
+	$(MKLIB) -o $(LIBNAME)nat $^ $(LDOPTS)
```
I know this comes from the old Makefile.nt, but is LDOPTS really useful (and if so, why not under Unix)?

alainfrisch replied:
Do we really want to build dllthreadsnat.dll?  This looks like a dll for bytecode, which seems useless here.  So one could use `$(MKLIB) -custom` (which becomes identical to the Unix version).

shindere replied:
Alain Frisch (2016/08/30 14:53 -0700):

> >  # Dynamic linking with -lpthread is risky on many platforms, so
> >  # do not create a shared object for libthreadsnat.
> > -libthreadsnat.a: $(NATIVECODE_C_OBJS)
> > -   $(AR) rc libthreadsnat.a $(NATIVECODE_C_OBJS)
> > -   $(AR) rc $@ $^
> >   +else # Windows
> > -   $(MKLIB) -o $(LIBNAME)nat $^ $(LDOPTS)
> 
> I know this comes from the old Makefile.nt, but is LDOPTS really
> useful (and if so, why not under Unix)?

No it's not useful, it's empty. Thanks!
Commit amended to not use it any longer.

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
 # do not create a shared object for libthreadsnat.
-libthreadsnat.a: $(NATIVECODE_C_OBJS)
-	$(AR) rc libthreadsnat.a $(NATIVECODE_C_OBJS)
+	$(AR) rc $@ $^
```
Open question: what about using `$(MKLIB) -custom`?  This brings it closer to the Windows version and possibly identical (see notes below).

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
-	$(CAMLOPT) -a -o threads.cmxa $(THREAD_OBJS:.cmo=.cmx) \
-	  -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
+$(LIBNAME).cmxa: $(THREADS_NCOBJS)
+	$(CAMLOPT) -a -o $@ $^ -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
```
I cannot match this with the previous behavior of Makefile.nt, which does not seem to pass any linker option to add libthreadsnat.lib when building threads.cmxa.  How did it work before, and do the new version change something?

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
+	  cp dllthreads$(EXT_DLL) $(INSTALL_STUBLIBDIR); fi
+	cp libthreads.$(A) $(INSTALL_LIBDIR)
+ifeq "$(UNIX_OR_WIN32)" "unix"
+	cd $(INSTALL_LIBDIR); $(RANLIB) libthreads.$(A)
```
It is fine to use $(RANLIB) under Windows to avoid the conditional (it is just defined as `echo`).

shindere replied:
Alain Frisch (2016/08/30 15:06 -0700):

> > -   if test -f dllthreads.so; then \
> > -     cp dllthreads.so $(INSTALL_STUBLIBDIR)/dllthreads.so; fi
> > -   cp libthreads.a $(INSTALL_LIBDIR)/libthreads.a
> > -   cd $(INSTALL_LIBDIR); $(RANLIB) libthreads.a
> > -   if test -d $(INSTALL_LIBDIR)/threads; then :; \
> > -     else mkdir $(INSTALL_LIBDIR)/threads; fi
> > -   cp $(THREAD_OBJS:.cmo=.cmi) threads.cma $(INSTALL_LIBDIR)/threads
> > -   rm -f $(INSTALL_LIBDIR)/threads/stdlib.cma
> > -   cp thread.mli mutex.mli condition.mli event.mli threadUnix.mli \
> > -      $(INSTALL_LIBDIR)
> > -   cp threads.h $(INSTALL_LIBDIR)/caml/threads.h
> > -   if test -f dllthreads$(EXT_DLL); then \
> > -     cp dllthreads$(EXT_DLL) $(INSTALL_STUBLIBDIR); fi
> > -   cp libthreads.$(A) $(INSTALL_LIBDIR)
> >   +ifeq "$(UNIX_OR_WIN32)" "unix"
> > -   cd $(INSTALL_LIBDIR); $(RANLIB) libthreads.$(A)
> 
> It is fine to use $(RANLIB) under Windows to avoid the conditional (it
> is just defined as `echo`).

Indeed, thanks!

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
-	cp thread.mli mutex.mli condition.mli event.mli threadUnix.mli \
-	   $(INSTALL_LIBDIR)
-	cp threads.h $(INSTALL_LIBDIR)/caml/threads.h
+	if test -f dllthreads$(EXT_DLL); then \
```
This would go away if we stop building dllthreads.dll (under Windows).

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
+endif
+	mkdir -p "$(THREADS_LIBDIR)"
+	cp $(CMIFILES) threads.cma "$(THREADS_LIBDIR)"
+	rm -f "$(THREADS_LIBDIR)/stdlib.cma"
```
Do you understand what this is about (I know it was there before)?

alainfrisch replied:
Yes:

```
rm -f "$(THREADS_LIBDIR)/stdlib.cma"
```

I don't know what is this stdlib.cma under `$(THREADS_LIBDIR)`.

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
+BYTECODE_C_OBJS=st_stubs_b.$(O)
+NATIVECODE_C_OBJS=st_stubs_n.$(O)
+
+THREADS_SOURCES := thread.ml mutex.ml condition.ml event.ml threadUnix.ml
```
Why are you using `:=` (and also below)?

shindere replied:
Alain Frisch (2016/08/30 15:10 -0700):

> >  COMPFLAGS=-w +33..39 -warn-error A -g -bin-annot -safe-string
> >  ifeq "$(FLAMBDA)" "true"
> >  OPTCOMPFLAGS=-O3
> >  else
> >  OPTCOMPFLAGS=
> >  endif
> > 
> > -BYTECODE_C_OBJS=st_stubs_b.o
> > -NATIVECODE_C_OBJS=st_stubs_n.o
> > +LIBNAME=threads
> > +
> > +ifeq "$(UNIX_OR_WIN32)" "unix"
> > +HEADER = st_posix.h
> > +else # Windows
> > +HEADER = st_win32.h
> > +endif
> > +
> > +BYTECODE_C_OBJS=st_stubs_b.$(O)
> > +NATIVECODE_C_OBJS=st_stubs_n.$(O)
> > +
> > +THREADS_SOURCES := thread.ml mutex.ml condition.ml event.ml threadUnix.ml
> 
> Why are you using `:=` (and also below)?

It's true that it may be better to be consistent.
That being said, I believe it is legitimate to use := in these places
and it would even be better to use it in other places, too.

See section 6.2 The Two Flavors of Variables in the make manual.

For the moment I replaced := by = everywhere, but I think in the long
run it may be better to use := where it makes sense, i.e. in a lot of
places, IMO.

alainfrisch commented:
Several simplifications could be added if we decide to drop the compilation of threads.cmxs and dllthreads.dll (both used to be built under Windows only). I'm in favor of doing it.

shindere commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
 
-libthreads.a: $(BYTECODE_C_OBJS)
-	$(MKLIB) -o threads $(BYTECODE_C_OBJS) $(PTHREAD_LINK)
+allopt: lib$(LIBNAME).$(A) $(LIBNAME).cmxa $(LIBNAME).cmxs $(CMIFILES)
```
Alain Frisch (2016/08/30 14:43 -0700):

> > -libthreads.a: $(BYTECODE_C_OBJS)
> > -   $(MKLIB) -o threads $(BYTECODE_C_OBJS) $(PTHREAD_LINK)
> >   +allopt: lib$(LIBNAME).$(A) $(LIBNAME).cmxa $(LIBNAME).cmxs $(CMIFILES)
> 
> This should be `lib$(LIBNAME)nat.$(A)`, no?

Well I'm getting confused. I believe that, at this level, the proposed
changes faithfully reflect the behaviour of the former Makefile.nt. Or
am I missing something here? Sorry if that's the case...

shindere commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
 # which itself will pass -lunix to the C linker.  It seems more
 # modular to me this way. -- Alain
 
+$(LIBNAME).cmxs: $(LIBNAME).cmxa lib$(LIBNAME)nat.$(A)
```
Alain Frisch (2016/08/30 14:47 -0700):

> The .cmxs file was not built on Unix previously, probably because
> "Dynamic linking with -lpthread is risky on many platforms" as another
> comment said.  It used to be built under Windows, though.  I'd rather
> not starting to build the .cmxs file under Unix, unless you have a
> good reason to do so.  And I would use the opportunity to stop
> building it under Windows as well for the sake of uniformity across
> platforms.

I think cmxs files are provided for other libraries, see for instance
the rules in otherlibs/Makefile. So my (open) question is: is it okay to
not be consistent in providing the .cmxs files?

I mean, as I understand it, there is a choice between either not
providing the .cmxs file for this library, for sake of obtaining a
simpler build system, or trying to provide the .cmxs file also for this
library, whenever possible (i.e. at least on Windows) because this is
more coherent with what happens with other libraries.

Is it correct to phrase the question that way?

I, personnally, have no strong opinion. It seems to me that the overall
coherence (providing .cmxs files whenever possible) could be considered as
more important than keeping the build system uniform, because it can
have an impact on users, whereas only OCaml developers are concerned
with the simplicity of the build system. But that beiing said I'm happy
to implement whatever the community finds best.

shindere commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,127 @@
+endif
+	mkdir -p "$(THREADS_LIBDIR)"
+	cp $(CMIFILES) threads.cma "$(THREADS_LIBDIR)"
+	rm -f "$(THREADS_LIBDIR)/stdlib.cma"
```
Alain Frisch (2016/08/30 15:08 -0700):

> > -   if test -d $(INSTALL_LIBDIR)/threads; then :; \
> > -     else mkdir $(INSTALL_LIBDIR)/threads; fi
> > -   cp $(THREAD_OBJS:.cmo=.cmi) threads.cma $(INSTALL_LIBDIR)/threads
> > -   rm -f $(INSTALL_LIBDIR)/threads/stdlib.cma
> > -   cp thread.mli mutex.mli condition.mli event.mli threadUnix.mli \
> > -      $(INSTALL_LIBDIR)
> > -   cp threads.h $(INSTALL_LIBDIR)/caml/threads.h
> > -   if test -f dllthreads$(EXT_DLL); then \
> > -     cp dllthreads$(EXT_DLL) $(INSTALL_STUBLIBDIR); fi
> > -   cp libthreads.$(A) $(INSTALL_LIBDIR)
> >   +ifeq "$(UNIX_OR_WIN32)" "unix"
> > -   cd $(INSTALL_LIBDIR); $(RANLIB) libthreads.$(A)
> >   +endif
> > -   mkdir -p "$(THREADS_LIBDIR)"
> > -   cp $(CMIFILES) threads.cma "$(THREADS_LIBDIR)"
> > -   rm -f "$(THREADS_LIBDIR)/stdlib.cma"
> 
> Do you understand what this is about (I know it was there before)?

Are you referring to one line in particular here?

shindere replied:
Alain Frisch (2016/08/31 08:46 -0700):

> Yes:
> 
> ```
> rm -f "$(THREADS_LIBDIR)/stdlib.cma"
> ```
> 
> I don't know what is this stdlib.cma under `$(THREADS_LIBDIR)`.

Me neither. Is it possible that, for older versions of OCaml, the stdlib
came in two flavours depending on whether one compiled with or without
threads? And then when newer versions arrived they added this command to
remove the thread-safe version of the stdlib which became obsolete?

Is that possible?

Perhaps this command can now be removed?

shindere commented:
Alain Frisch (2016/08/30 14:55 -0700):

> >  # Dynamic linking with -lpthread is risky on many platforms, so
> >  # do not create a shared object for libthreadsnat.
> > -libthreadsnat.a: $(NATIVECODE_C_OBJS)
> > -   $(AR) rc libthreadsnat.a $(NATIVECODE_C_OBJS)
> > -   $(AR) rc $@ $^
> 
> Open question: what about using `$(MKLIB) -custom`?  This brings it
> closer to the Windows version (and possibly identical if one decide
> not to build libthreadsnat.dll anymore).

Done.

Alain Frisch (2016/08/30 14:56 -0700):

> >  # Dynamic linking with -lpthread is risky on many platforms, so
> >  # do not create a shared object for libthreadsnat.
> > -libthreadsnat.a: $(NATIVECODE_C_OBJS)
> > -   $(AR) rc libthreadsnat.a $(NATIVECODE_C_OBJS)
> > -   $(AR) rc $@ $^
> >   +else # Windows
> > -   $(MKLIB) -o $(LIBNAME)nat $^ $(LDOPTS)
> 
> Do we really want to build dllthreadsnat.dll?  This looks like a dll
> for bytecode, which seems useless here.  So one could use `$(MKLIB)
> -custom` (which becomes identical to the Unix version).

So we discussed with @damiendoligez. He also thinks building the shared
objects is not useful. Hence the new version of the commit which does
not build them any longer.

Alain Frisch (2016/08/30 15:07 -0700):

> >  INSTALL_STUBLIBDIR=$(DESTDIR)$(STUBLIBDIR)
> > 
> >  install:
> > -   if test -f dllthreads.so; then \
> > -     cp dllthreads.so $(INSTALL_STUBLIBDIR)/dllthreads.so; fi
> > -   cp libthreads.a $(INSTALL_LIBDIR)/libthreads.a
> > -   cd $(INSTALL_LIBDIR); $(RANLIB) libthreads.a
> > -   if test -d $(INSTALL_LIBDIR)/threads; then :; \
> > -     else mkdir $(INSTALL_LIBDIR)/threads; fi
> > -   cp $(THREAD_OBJS:.cmo=.cmi) threads.cma $(INSTALL_LIBDIR)/threads
> > -   rm -f $(INSTALL_LIBDIR)/threads/stdlib.cma
> > -   cp thread.mli mutex.mli condition.mli event.mli threadUnix.mli \
> > -      $(INSTALL_LIBDIR)
> > -   cp threads.h $(INSTALL_LIBDIR)/caml/threads.h
> > -   if test -f dllthreads$(EXT_DLL); then \
> 
> This would go away if we stop building dllthreads.dll (under Windows).

It has indeed been removed.

Alain Frisch (2016/08/30 15:13 -0700):

> Several simplifications could be added if we decide to drop the
> compilation of threads.cmxs and dllthreads.dll (both used to be built
> under Windows only). I'm in favor of doing it.

Done. If you see other simplifications your comments are welcome.

(making make -j work in this directory seems to be a different matter
for another PR, to keep this one of a reasonable size)

Alain Frisch (2016/08/31 03:15 -0700):

> > -libthreads.a: $(BYTECODE_C_OBJS)
> > -   $(MKLIB) -o threads $(BYTECODE_C_OBJS) $(PTHREAD_LINK)
> >   +allopt: lib$(LIBNAME).$(A) $(LIBNAME).cmxa $(LIBNAME).cmxs $(CMIFILES)
> 
> Yes, but it was probably wrong (and Makefile was right).

Fixed.

Alain Frisch (2016/08/31 03:20 -0700):

> >  # Note: I removed "-cclib -lunix" from the line above.
> >  # Indeed, if we link threads.cmxa, then we must also link unix.cmxa,
> >  # which itself will pass -lunix to the C linker.  It seems more
> >  # modular to me this way. -- Alain
> > 
> > +$(LIBNAME).cmxs: $(LIBNAME).cmxa lib$(LIBNAME)nat.$(A)
> 
> Simplifying the build system is secondary, I agree.  But even for
> users, declaring once and for all (and documenting) that the
> systhreads library does not provide a .cmxs file is better, in my
> opinion, than providing the .cmxs file only on Windows.

@dra27 you may want to review this PR.

A few additional remarks:
- It seems the use of -linkall under Windows has been there for a while,
  and had been introduced in a subversion branch called natdynlink,
  presumably by you @alainfrisch. Do you remeber what lead you to
  introduce this?

In case you do not remember, we (@damiendoligez and I) propose to use
-linkall everywhere because the linked modules are not that big and, for
most of them (except perhaps events which is really small) it does not
seem to make sense to use one without the others.

I also removed two things:
- The dependency of .cmx files on the ocamlopt comiler, because
  it is not there in other libraries and because it had been added
  almost 20 years ago, presumably for debugging purposes.
- The line that removes threads/stdlib.cma, in agreement with
  @damiendoligez.

alainfrisch commented:
> -linkall under Windows has been there for a while, and had been introduced in a subversion branch called natdynlink, presumably by you @alainfrisch. Do you remeber what lead you to introduce this?

No, unfortunately, I don't remember (that branch was the one were the notion of .cmxs plugins was added).

shindere commented:
Alain Frisch (2016/09/05 08:21 -0700):

> > -linkall under Windows has been there for a while, and had been introduced in a subversion branch called natdynlink, presumably by you @alainfrisch. Do you remeber what lead you to introduce this?
> 
> No, unfortunately, I don't remember (that branch was the one were the
> notion of .cmxs plugins was added).

Okay. So I can add the -linkall on the Unix side as suggested earlier.

This, however, leads to two other questions:
1. Even if -linkall is added, the commands on Unix and Windows
   will still be different, because on the Unix side we have
   -cclib -lunix $(PTHREAD_CAML_LINK)
   which is not there on the Windows side.
   (To be complete, $(PTHREAD_CAML_LINK) comes from configure, see
   configure variable $pthread_caml_link.)

Can anyone see a way to unify these two commands?
1. If threads.cma is built using -linkall everywhere, shouldn't
   threads.cmxa be built with -linkall, too?

shindere committed:
Make sure the PTHREAD_CAML_LINK variable is defined everywhere.

shindere commented:
The PR has been rebased on latest trunk and updated so that -linkall is
used consistently.

Moreover, comments have been added about those points that need further
clarification.

Can anybody see anything that still needs to be done before merging?

Thanks.

alainfrisch commented:
The CI build fails with:

```
../../boot/ocamlrun ../../ocamlopt -nostdlib -I ../../stdlib -I ../../otherlibs/unix -a -o -linkall threads.cmxa thread.cmx mutex.cmx condition.cmx event.cmx threadUnix.cmx -cclib -lthreadsnat -cclib -lpthread
Option -a cannot be used with .cmxa input files.
```

(-linkall should not be added between -o and its argument)

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,107 @@
-	$(CAMLOPT) -a -o threads.cmxa $(THREAD_OBJS:.cmo=.cmx) \
-	  -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
+$(LIBNAME).cmxa: $(THREADS_NCOBJS)
+	$(CAMLOPT) -a -o -linkall $@ $^ -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
```
This line causes the build failure.

shindere replied:
Fixed. Sorry for the noise, thanks a lot for your prompt report,
@alainfrisch!

alainfrisch commented:
AppVeyor fails with

```
cp libthreads.lib C:/Program Files/OCaml/lib
cp: target 'Files/OCaml/lib' is not a directory
```

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,107 @@
-	cp thread.mli mutex.mli condition.mli event.mli threadUnix.mli \
-	   $(INSTALL_LIBDIR)
-	cp threads.h $(INSTALL_LIBDIR)/caml/threads.h
+	cp libthreads.$(A) $(INSTALL_LIBDIR)
```
Missing quotes around the target directory, I guess.

shindere replied:
Alain Frisch (2016/09/09 06:18 -0700):

> AppVeyor fails with
> 
> ```
> cp libthreads.lib C:/Program Files/OCaml/lib
> cp: target 'Files/OCaml/lib' is not a directory
> ```

Alain Frisch (2016/09/09 06:19 -0700):

> >  INSTALL_STUBLIBDIR=$(DESTDIR)$(STUBLIBDIR)
> > 
> >  install:
> > -   if test -f dllthreads.so; then \
> > -     cp dllthreads.so $(INSTALL_STUBLIBDIR)/dllthreads.so; fi
> > -   cp libthreads.a $(INSTALL_LIBDIR)/libthreads.a
> > -   cd $(INSTALL_LIBDIR); $(RANLIB) libthreads.a
> > -   if test -d $(INSTALL_LIBDIR)/threads; then :; \
> > -     else mkdir $(INSTALL_LIBDIR)/threads; fi
> > -   cp $(THREAD_OBJS:.cmo=.cmi) threads.cma $(INSTALL_LIBDIR)/threads
> > -   rm -f $(INSTALL_LIBDIR)/threads/stdlib.cma
> > -   cp thread.mli mutex.mli condition.mli event.mli threadUnix.mli \
> > -      $(INSTALL_LIBDIR)
> > -   cp threads.h $(INSTALL_LIBDIR)/caml/threads.h
> > -   cp libthreads.$(A) $(INSTALL_LIBDIR)
> 
> Missing quotes around the target directory, I guess.

Alain Frisch (2016/09/09 06:20 -0700):

> >  installopt:
> > -   cp libthreadsnat.a $(INSTALL_LIBDIR)/libthreadsnat.a
> > -   cd $(INSTALL_LIBDIR); $(RANLIB) libthreadsnat.a
> > -   cp $(THREAD_OBJS:.cmo=.cmx) threads.cmxa threads.a \
> > -      $(INSTALL_LIBDIR)/threads
> > -   cd $(INSTALL_LIBDIR)/threads && $(RANLIB) threads.a
> > -   cp libthreadsnat.$(A) "$(INSTALL_LIBDIR)"
> > -   cd $(INSTALL_LIBDIR); $(RANLIB) libthreadsnat.$(A)
> > -   cp $(THREADS_NCOBJS) threads.cmxa threads.$(A) $(THREADS_LIBDIR)
> 
> Missing quotes as well (also on `cd` lines above and below).

Indeed, sorry about that and thanks a lot for having reported it.

This should now be fixed and the PR has been rebased on the latest
trunk.

alainfrisch commented on otherlibs/systhreads/Makefile:
```diff
@@ -13,96 +13,107 @@
-	cd $(INSTALL_LIBDIR)/threads && $(RANLIB) threads.a
+	cp libthreadsnat.$(A) "$(INSTALL_LIBDIR)"
+	cd $(INSTALL_LIBDIR); $(RANLIB) libthreadsnat.$(A)
+	cp $(THREADS_NCOBJS) threads.cmxa threads.$(A) $(THREADS_LIBDIR)
```
Missing quotes as well (also on `cd` lines above and below).

alainfrisch commented:
Thanks again for your hard work Sébastien!

I suggest that we group all these related changes into a single entry in the Changes file, if you agree.

shindere commented:
Alain Frisch (2016/09/13 08:59 -0700):

> Merged #785.

Thanks for merging @alainfrisch, and thanks even more for your patience
in reviewing.

Hope things will be easier with #808.

shindere commented:
Alain Frisch (2016/09/13 09:00 -0700):

> Thanks again for your hard work Sébastien!

Well, sorry for the lack of rigor!

> I suggest that we group all these related changes into a single entry
> in the Changes file, if you agree.

Certainly!

ghost commented:
I'm getting some linking errors with 4.05 as dllthreads.so is no longer installed. This seems to be caused by this PR. Is it intentional?

ghost commented:
Note that ocamlrun is linked with -lpthread, so dynamic loading of threads.cma doesn't cause dynamic linking of -lpthread

dra27 commented:
It looks like an accident - "both" platforms.

dra27 commented:
It should be `dllthreads.$(EXT_DLL)` now - are you able to do a quick GPR for it?

dra27 commented:
We really should alter the testsuite and CI so that the testsuite is run over an installed OCaml, not just a compiled one!

ghost commented:
Agreed for the testsuite. I think the fix can be committed directly. Should `dllthreads.$(EXT_DLL)` be installed on Windows as well?

dra27 commented:
Yes to everything!

ghost commented:
Ok, fixed in 812eb68de964b59deed6aba06d2a4feb0043c250

shindere commented:
Jérémie Dimino (2017/01/27 09:50 -0800):
> Ok, fixed in 812eb68de964b59deed6aba06d2a4feb0043c250

Many thanks Jérémie and David for your work on this! Sorry for the
error!!

Agreed that the testsuit should work against an installed version of
OCaml. Took note of this and will take it into account!

gasche commented:
Being able to run the testsuite from the compiled checkout without installation is important to me. The reason is that I have to configure a source checkout carefully to be able to install it as an opam switch, and I don't always remember to do this when I just run a build to test something. Forcing me to install would either force me to reconfigure (and rebuild) or risk overwriting the currently-configured switch location.

Having both test-from-sources and test-from-install targets would of course be fine.

dra27 commented:
@gasche - I completely agree, I think the ability to do *both* is important! Having done that, we could also configure the two Travis tests to use the both approaches (e.g. flambda installs, non-flambda doesn't).

@shindere - thanks, I'll have a look. Get well soon - hope that arm's mending!

shindere commented:
Gabriel Scherer (2017/01/28 03:00 -0800):
> Having both test-from-sources and test-from-install targets would of
> course be fine.

That's what I'd like to achieve.

test-from-install seems especially important in the context of
cross-compiling.

shindere commented:
David Allsopp (2017/01/28 03:13 -0800):
> @shindere - thanks, I'll have a look. Get well soon - hope that arm's
> mending!

Thanks. I'm a bit worried because it seems to evolve very slowly, if at
all. Rather frustrating! :)

ghost commented:
@shindere no poblem, get well soon. I'm glad we can now build all of JS packages against trunk on a daily basis. This should help as well with testing trunk

dra27 commented:
@diml - actually, https://github.com/ocaml/ocaml/commit/fbdddb2205aaf5a7c20c7f4cec06ebb74094e5a3 suggests that maybe we should never push build system changes directly :)

ghost commented:
@dra27, I disagree, I think it suggests that `bash` is a terrible tool to use in build systems :)

dra27 commented:
:) After my recent battles with AppVeyor in #1024, I couldn't agree more with that! As soon as you want spaces in filenames and directories, the same becomes true for make!
